### ISSUE TITLE
Segmentation fault in fgmres-ams solver when using mfem-complex-mfem-dev

### ISSUE THREAD

**Issue by qiuchangkai** [mfem/mfem#632]
Dear Mark @mlstowell and other developers,

I am using mfem-complex-mfem-dev and hypre-2.11.2 to solve the complex-value Maxwell's equation

1/mu curl curl E + i omega sigma E = - i omega Js

The code are modified from ex3p and ex21p_proposed, where I use my own mesh files. 

The program reports signal SIGSEGV, Segmentation fault after first fgmres iteration 
for some input grids

Program received signal SIGSEGV, Segmentation fault.
0x00000000007d6094 in hypre_AMSComputePixyz (A=0x349ed80, G=0x572fb20, Gx=0x292e7b0, Gy=0x292e8c0, Gz=0x292e930, dim=3, Pix_ptr=0x349f0c8, Piy_ptr=0x349f0d0, Piz_ptr=0x349f0d8) at ams.c:1714
1714	               *Pix_offd_data++ = fabs(G_offd_data[j]) * 0.5 * Gx_data[i];

I find that if the function fespace->GetEssentialTrueDofs(ess_bdr, ess_tdof_list) is not called, this error won't happen.

In my mesh file, the triangles on the outer boundary of the tetrahedron grid are marked as the boundary.

I didn't set up the linear form like the examples but calculated the final vector by myself, because I don't know how I can define the vector delta coefficients Js in some elements for my application. I am wondering whether this will affect the solver.

It's beyond my ability to step into ams to see what happened. But I am very glad to provide you with the code and input files to debug this issue.

Thank you very much for your kind help!

Best,
Changkai

**Comment by qiuchangkai** [comment]
I didn't set up the linear form like the examples but calculate the final vector by myself, because I don't know how I can define the vector delta coefficients Js in some elements.

**Comment by mlstowell** [comment]
Hello, @qiuchangkai ,
I've just tried `ex21p_proposed` with a tetrahedral mesh and it ran without any issues but there could still be a problem.  I believe this example solves:
1/mu curl curl E - omega^2 epsilon E + i omega sigma E = - i omega Js
This has one additional mass matrix added to the real part of the operator which makes the linear system non-singular.  It's possible that without this mass matrix AMS is unable to solve your linear system but this would probably lead to a floating point exception rather than a segmentation fault.  
Are you passing `1/mu curl curl + omega sigma` as the operator to AMS?  Is your `sigma` nonzero everywhere in your problem domain?
It may also be a good idea to try a sparse direct solver like SuperLU, STRUMPACK, or MUMPs if you can.
Can you verify that your mesh can be used with an unmodified version of `ex21p_proposed`?  I doubt this is a mesh related problem but this test should confirm that assumption.
Best wishes,
Mark

**Comment by qiuchangkai** [comment]
Hi Mark, @mlstowell 
Thanks for your prompt reply. I only add the term i omega sigma E.  Some codes are like:
        
        PWConstCoefficient sigma(lambda);
	pcOp->AddDomainIntegrator(new CurlCurlIntegrator(*muinv));
	pcOp->AddDomainIntegrator(new VectorFEMassIntegrator(sigma));
	ParSesquilinearForm *ac = new ParSesquilinearForm(fespace, conv);
	ac->AddDomainIntegrator(new CurlCurlIntegrator(*muinv),NULL);
	ac->AddDomainIntegrator(NULL,new VectorFEMassIntegrator(sigma));
        ComplexHypreParMatrix * Ahyp = dynamic_cast<ComplexHypreParMatrix*>(AC.Ptr());
	
	pc_r = new HypreAMS(dynamic_cast<HypreParMatrix&>(*PCOp.Ptr()),fespace);
	pc_i = new ScaledOperator(pc_r,(conv == ComplexOperator::HERMITIAN) ?1.0:-1.0);

I am passing 1/mu curl curl + omega sigma as a operator to AMS. It's a geophysical electromagnetic problem, where some elements are for air,  and I set their conductivity to 1.0e-8. Sigma is nonzero everywhere, but very small in some elements.

The original ex21p_proposed source code works well for my own mesh.

I need some time to modify the code to try MUMPS from PETSc. I will update asap.

Changkai

**Comment by mlstowell** [comment]
Hi, @qiuchangkai ,
It sounds like you're doing everything correctly.  However, you're probably the first person to really exercise the new classes in that particular branch of MFEM so I'm not surprised you've run into difficulty (but I am sorry about that).
Are you familiar with this paper https://library.seg.org/doi/10.1190/geo2015-0013.1?  They seem to be solving the same linear system that you are and they also use AMS.  It may contain some useful advice on applying AMS in this area.
Best wishes,
Mark

**Comment by tzanio** [comment]
Hi @qiuchangkai,

Can you reproduce this problem in example 3p (e.g. with `i omega sigma` replaced by `sigma`)?

The line you get the segfault in seems to indicate an issue with the discrete gradient matrix `G`, which should not have the boundary conditions eliminated, but that's something we should be handling correctly internally...

Can you confirm also that you are running 1st order elements? What happens if you increase the order?

Cheers,
Tzanio

**Comment by qiuchangkai** [comment]
Dear Mark @mlstowell ,

I read Alexander and Tzanio's  paper years ago. Actually their paper inspired me to use fgmres-ams to solve the complex-value Maxwell's equation with tetrahedron grids. I implemented my own FEM code with direct solver, which is expensive in memory usage.

**Comment by qiuchangkai** [comment]
Dear Mark @mlstowell ,

I read Alexander and Tzanio's  paper years ago. Actually their paper inspired me to use fgmres-ams to solve the complex-value Maxwell's equation with tetrahedron grids. I implemented my own FEM code with direct solver, which is expensive in memory usage.

**Comment by qiuchangkai** [comment]
Hi Tzanio @tzanio , the original example 3p also works fine for my grid. I am running first order elements. If I increase the order, the problem would not happen. 

Today I find that it may be because of the code I wrote to calculate the current source Js.  In summary, ex3p and ex21p_proposed would work well for my grid and piece wise coefficient sigma.  But if the code calculating the source term Js is called, this segmentation fault would occur. 

I call some member functions of the class pmesh/mesh to find the geometry relationship for the grids.  Will this alter the internal data structure of the class pmesh, and further affect the boundary conditions and the construction of  the discrete gradient matrix G? 

I am still debugging the program...

Best,
Changkai

**Comment by qiuchangkai** [comment]
Hi Tzanio and Mark, I finally make the modified code run sucessfully but don't know if the solution is correct. 

I want to ask another question about how to calculate the electromagnetic field (i.e., calculate the shape functions) at some physical points. I prefer to do this manually rather than using the class GridFunction because the shape functions and curl of shape functions are needed for other purposes. 

Given a point with physical coordinates [x,y,z], I can find an element with global index ielement which contains this point. What I do is like:

		const FiniteElement *FElem = fespace->GetFE(ielement);
		DenseMatrix vshape(dof, spaceDim);
		DenseMatrix curlshape(dof, spaceDim);
		ElementTransformation *Tr = fespace->GetElementTransformation(ielement);
		Vector pt;
		pt.SetSize(3);

		pt[0] = x; // What I should do here. I'm rather confused.
		pt[1] = y;
		pt[2] = z;
		IntegrationPoint ip;// What I should do here. I'm rather confused.
		Tr->TransformBack(pt, ip); // What I should do here. I'm rather confused.
		Tr->SetIntPoint(&ip);// What I should do here. I'm rather confused.
		// Calculate the shape functions
		FElem->CalcVShape(*Tr, vshape);
		// Calculate the curl of shape functions
		FElem->CalcPhysCurlShape(*Tr, curlshape);

		Array<int> vdofs;
		fespace->GetElementVDofs(ielement, vdofs);
		// Then I can perform the product of basis functions, the curl of the basis functions with the solution.

Is the above code correct? Are the returned shape and curl functions all in the physical space?

Thanks again.
Changkai

**Comment by mlstowell** [comment]
Hi, @qiuchangkai ,

Yes, that looks correct to me.

Best wishes,
Mark

**Comment by qiuchangkai** [comment]
Hi Mark @mlstowell , the code looks correct. However, the final electric field at some interested physical points are not correct. Please see the attached figure for details. 
![Uploading untitled.jpg…]() There are some jump points which are wrong, while others are very close with existing results. The results for magnetic field are worse. I just infer there is no problem with the solution by the gmres solver. I can confirm that the global elements which contains these physical points are found correctly. Actually, they are one of the vertices in the physical elements, so after the transformation the coordinates of the integral points are either 0 or 1. How can I improve the results like these?

Best,
Changkai

**Comment by mlstowell** [comment]
Hi, @qiuchangkai ,

You probably already realize this but I wanted to be sure.  When you call `fespace->GetElementVDofs(ielement, vdofs);` some of the entries in `vdofs` can be negative for Nedelec and Raviart-Thomas bases.  This indicates that the orientation of the corresponding basis function should be reversed.  The best way to handle this is to get the local degrees of freedom using `gridfunc->GetSubVector(vdofs, elemvect);` which processes these negative entries correctly (if you're curious the implementation  is in `Vector::GetSubVector()`).

Best wishes,
Mark

**Comment by TangLaoya** [comment]
Dear Changkai,
How do you process current source in vector finite element? I am trying to solve induction logging problem, I need to calculate the induced voltage of received coils from source coils, my problem is how to set up current source of source coils for the term 
b=-i*omega*int(dot(N__i, J))

Could you please give me some clue on how to fill right hand side after the discretization of domain into tetrahedron elements?

Thanks,
Tang Laoya

**Comment by qiuchangkai** [comment]
Dear Tang Laoya @TangLaoya ,

What I am doing is dividing any transmitting sources (include a current loop and a current wire) into a series of short electric dipoles. Then I can calculate the small electric dipoles (its center, direction, length) inside each elements which contains the source. Actually, this step is the main work. At last, the small electric dipole can be expressed as a vector delta function, so I can compute the integral of the basis functions and current source Js directly using the property of delta functions. Remember to refine the elements near the source, thus each small segments of source can be treated as dipoles. Hope this would help.

Best,
Changkai

**Comment by mlstowell** [comment]
HI, @TangLaoya ,

What you are describing is precisely what the `(Par)LinearForm` objects are used for.  I assume you can write your current density as a function of position so you can implement:
```void JFunc(const Vector &x, Vector &J);```
Then  you define an MFEM `VectorCoefficient` using this function:
```VectorFunctionCoefficient JCoef(dim, JFunc);```
If `JFunc` cannot be written then you may need to use a `VectorGridFunctionCoefficient` instead (assuming you have a `GridFunction` representation of `J`.)
Then setup the `LinearForm`:
```
LinearForm J(fespace);
J.AddDomainIntegrator(new VectorFEDomainLFIntegrator(JCoef));
J.Assemble(); /// This step performs the integration you mentioned
```
You can look at examples `ex3` and `ex3p` or the `tesla` miniapp for more details.  This scheme assumes that `JFunc` will be evaluated everywhere in the problem domain so it would need to return the zero vector outside of your rectangular region.  This is probably easy enough for a rectangular region but in some cases the region is too complicated.  An alternative is to mark the region containing the current density using a separate volume attribute.  The use a `VectorRestrictedCoefficient` which will only be applied to elements with this attribute.

Best wishes,
Mark

**Comment by TangLaoya** [comment]
Dear Mark,
Thank you very much for your kindly reply. The example 'tesla' is just what I am looking for. But there is still a problem for me: the hypre solver can't be linked into the MFEM, I have build the hypre by CMakeLists.txt by Cmake in Windows 7 64bit + VS 2013, but it can't be linked to the MFEM, I have reported this problem before, could anyone help me to take a look at it?

Thanks,
Tang Laoya

**Comment by mlstowell** [comment]
Hi, @TangLaoya ,

Are you referring to issue #620?  If so, I believe we do have a solution, see PR #633.

Mark

**Comment by TangLaoya** [comment]
No, the issue #620 is solved. I mean the issue #621, in which I said the hypre can't be linked into MFEM, so I suggested to provide a serial version of hertz. 

Thanks,
Tang Laoya

**Comment by mlstowell** [comment]
Oh, okay.  I thought that was resolved as well but you are correct we only found a workaround.  We should continue discussing this in that issue.

**Comment by TangLaoya** [comment]
Dear Mark,
Thank you very much for your kindly reply. The steps I build hypre and MFEM:
1) put hypre to 
E:\mfem\hypre2.11.2
2) build hypre by Cmake
  --set 'where is the source code:' to 'E:\mfem\hypre2.11.2\src', set 'where to build the binaries:' to 'E:\mfem\hypre2.11.2\src\64bit_VS2013', and the configure, generate the project;
  --buid hypre by VS2013, passed
3) configure and build MFEM
  --fill 'HYPRE->HYPRE_DIR' to 'E:\mfem\hypre2.11.2\src\64bit_VS2013'
  --fill 'HYPRE->HYPRE_INCLUDE_DIR' to 'E:\mfem\hypre2.11.2\src'
  --fill 'HYPRE->HYPRE_INCLUDE_DIRS' to 'E:\mfem\hypre2.11.2\src'
  --fill 'HYPRE->HYPRE_LIBRARIES' to 'E:\mfem\hypre2.11.2\src\lib'
  --fill 'HYPRE->HYPRE_LIBRARY' to 'E:\mfem\hypre2.11.2\src\lib'
  --select 'MFEM->MFEM_USE_MPI', and other options such build examples.
  then configure and generate, the following error message displayed:
CMake Error at config/cmake/modules/FindHYPRE.cmake:33 (message):
  Unable to determine HYPRE version.
Call Stack (most recent call first):
  CMakeLists.txt:131 (find_package)

Thanks,
Tang Laoya

**Comment by v-dobrev** [comment]
Try setting `HYPRE_LIBRARIES` to the full path to the file `HYPRE.lib`, including `HYPRE.lib` -- that should be name of the hypre library built in step 2.

Also, you probably need to set `HYPRE_INCLUDE_DIRS` to point to the hypre `include` directory - this is where files like `_hypre_parcsr_mv.h` must be located.

The rest of the `HYPRE*` variables that you are setting should not be needed.

**Comment by mlstowell** [comment]
Hello, @qiuchangkai ,

Perhaps the easiest way to find this information for lowest order Nedelec elements is to build a discrete gradient operator:
```
   grad = new ParDiscreteLinearOperator(&HGradFESpace, &HCurlFESpace);
   grad->AddDomainInterpolator(new GradientInterpolator());
   grad->Assemble();
```
This code was taken from `miniapps/electromagnetics/joule_solver.cpp`.  Each row of this operator corresponds to a local Nedelec degree fo freedom, which lie on edges, and the columns correspond to the vertices which terminate these edges.  The values of the operator will be plus or minus one; -1 for the tail and +1 for the head of the vector.  For high order basis functions many of the Nedelec degrees of freedom don't correspond to edges so their vectors don't terminate on vertices.

I'm curious why you need this information.  Perhaps we can offer a more convenient way to accomplish your goal.

Best wishes,
Mark

**Comment by mlstowell** [comment]
Hi, @qiuchangkai ,

If I understand you correctly you're solving a Laplace problem for the electric potential as part of your initial condition.  If you're using MFEM for this solve you probably compute the electric potential as an H1 `GridFunction`.  To compute the initial electric field from this you can simply use the `DiscreteGradOperator` similar to what's done in the `volta` miniapp (or the code shown above from the `joule` miniapp which is essentially the same).  In either case the relationships between the H1 and H(Curl) dofs within a particular element are computed by the `GradientInterpolator` for arbitrary order basis functions.  The `DiscreteLinearOperator` then uses the `FiniteElementSpaces` to correctly handle the `vdof` orientations and copy the local element matrices into a global linear operator.  You may not need to manipulate the `vdofs` directly unless you're doing something more exotic.

Best wishes,
Mark

**Comment by qiuchangkai** [comment]
Hi Mark @mlstowell , could you please see line 386 in source code volta_solver.cpp? Should the comment "magnetic field" be "electric field", i.e., `grad*phi=-e` , but not `grad*phi=-h`? Do the electric field computed from the scalar potential keep the same directions with all the `vdofs` in H(curl) space? I am wondering whether I have to adjust the directions of the electric field from `grad*phi=-e` when I want to perform vector operatitons with the electric field from the H(curl) space.

Best,
Changkai

**Comment by mlstowell** [comment]
Hi, @qiuchangkai ,

The comment on line 386 should indeed refer to the electric field.  Thank you for bringing this to our attention.  I'll correct it.

The short answer to your question is "no" you do not need to do anything to adjust `e` after computing it as `-grad*phi`.  The reason for this is that the H(curl) space contains the gradients of the basis functions for the H_1 space.  The consequence of this is that the degrees of freedom for a gradient field in H(curl) are a simple linear combination of the degrees of freedom for a field in H_1 and the `grad` operator simply contains the appropriate linear combinations.

The `GetSubVector` method is actually a linear algebra method and not a finite element method.  This means that it does nothing particularly associated with H(curl) or physical space since these are finite element concepts not linear algebra concepts.  The `GetSubVector` method simply returns a collection of entries from a vector that are specified by an array of indices (this array is called `vdofs` in this context but they could be any set of indices).  When `GetSubVector` is used within `GridFunction` objects the degrees of freedom it returns are combined with their corresponding basis functions to compute the electric field, in this case.   It's the basis functions, not the entries returned through `GetSubVector`, that would need to be adjusted to properly align with physical space.  This alignment is handled by the `FiniteElement` classes through methods like `CalcPhysShape` or `CalcPhysVShape`.

Best wishes,
Mark

**Comment by qiuchangkai** [comment]
Hi @mlstowell ,
Thanks for your kind help and explanation. I will close this issue soon. By the way, what is the best way to ask some general questions about the usages of mfem in the future?

Best,
Changkai

**Comment by mlstowell** [comment]
Hi, @qiuchangkai ,

I think this is the best forum for such questions.  It gives all of our team members a chance to read the questions and keep abreast of the concerns of our users.

Best wishes,
Mark

**Comment by qiuchangkai** [comment]
Okay. Thanks.

**Comment by qiuchangkai** [comment]
Hi @mlstowell , I reopen this topic to ask a question about how to calculate the divergence of the current density, i.e, the operator `Div(sigma E)`, where sigma is the `PWConstCoefficient,` and `E` is a known vector from the Nedelec fespace. 

I first calculate the mass matrix `ParBilinearForm->AddDomainIntegrator(new VectorFEMassIntegrator(sigma))` , then multiply mass matrix with vector `E` to obtain the current density `J.`

The question is how to calculate the divergence of the Vector `J`.  I found that, in miniapps, there is a  `ParMixedBilinearForm * weakDiv_'  in the class IrrotationalProjector` . Is this `wearkDiv_` operator what I am looking for? 

What does the parameter  `const int & irOrder` mean?

**Comment by qiuchangkai** [comment]
Hi @mlstowell , I also want to report a problem about the `FGMRESSolver`.

The problem happens when users want to run `FGMRESSolver` without any restart.  The following code would expect the FGMRESSolver stops when exceeds the maximum number of iterations or converges, but without any restart.

`fgmres->SetKDim(a_very_large_number);`
`fgmres->SetMaxIter(maxit);`

In practice, the `FGMRESSolver` will restart several times after `maxit` iterations. It may be caused by the counter `j++`, please see line 830 in the source code `solvers.cpp`. This counter is not updated after each iteration when FGMRESSolver is not restarted. Moving `j++` to line 790 could partially solve the problem, but the `FGMRESSolver` would still restart and then return. The solver should return the solution after `maxit` iterations without restart in this case.

Hope I understand the code correctly.

Best,
Changkai

**Comment by mlstowell** [comment]
Hi, @qiuchangkai ,
The `irOrder` parameter of the `IrrotationalProjector` class is the integration rule order to be used for the bilinear forms inside this object.  This is needed so that the integration rules are compatible with the `CurlCurlIntegrator` that is presumed to be created outside of this object.  The goal is that the vectors produced by `IrrotationalProjector` should satisfy `S*v=0` where `S` is a stiffness matrix based on `CurlCurlIntegrator`.

Now regarding the divergence of a Nedelec field.  There are at least three ways to compute this and each has pros and cons.  The first two are based on the `VectorFEWeakDivergenceIntegrator` used in `IrrotationalProjector`.  I can go into the details of each but that's a rather long post.  Let me summarize each and we can work out the details if needed.

Option 1) Compute `Div(J)` as a field in H_1 using `E`, in H(Curl), and the coefficient `sigma`.  To do this you multiply the Nedelec grid function representing `E` by a mixed bilinear form built with `VectorFEWeakDivergenceIntegrator(sigma)` or `MixedVectorWeakDivergenceIntegrator(sigma)` (the second one supports tensor valued `sigma`). You then construct and solve an H_1 mass matrix using the product you just computed as the right hand side.  Pros: single matvec and solve. Cons: no intermediate `J`, `Div(J)` can be inaccurate near boundary.

Option 2) Compute `Div(J)` as a field in H_1 using `J`, in H(Curl), as an intermediate field.  To do this you proceed, as you were thinking, with a `VectorFEMassMatrix(sigma)`.  This requires that you solve a Nedelec mass matrix, with coefficient 1, to compute `J` from the `sigma E` which you computed using `VectorFEMassMatrix(sigma)` and `E`.  You then proceed to solve for `DivJ` as in option 1 except that your weak divergence operator now uses 1 for its coefficient. Pros: intermediate `J`.  Cons: 2 mavtecs and 2 solves, tangential rather than normal continuity for `J` which is non-physical, `Div(J)` can be inaccurate near boundary.

Option 3) Compute `Div(J)` as a field in L^2 using `J` in H(Div) as an intermediate field.  This is similar to option 2 except that you use `VectorFEMassMatrix(sigma)` in a `ParMixedBilinearForm` and then solve for `J` using an RT mass matrix.  The divergence of `J` is then computed using a `ParDiscreteDivOperator` which does not require a second solve. Pros: proper continuity for `J`, reliable values for `Div(J)` everywhere, together these make charge conservation easier to maintain.  Cons: one additional matvec compared to option 1.

I have a strong preference for option 3 because it produces a `J` which has the correct type of continuity at material interfaces.  The other issue with options 1 and 2 is the implicit boundary condition that is applied via the weak divergence operator which can lead to non-physical divergence values at boundary nodes.

Best wishes,
Mark

**Comment by mlstowell** [comment]
Hi, @qiuchangkai ,

I'll take a look at our FGMRES implementation.  I seem to recall seeing odd behavior as well.

Thanks,
Mark

**Comment by mlstowell** [comment]
Hi, @qiuchangkai ,

It looks like MFEM's implementations of GMRES and FGMRES use different notions of "Iteration" and "Pass".  It seems to me that the GMRES notion makes more sense.  I'll try to put together a pull request which brings these into conformity and we'll see what the other developers think.

Incidentally, restarted GMRES methods do assume that the Krylov space dimension is "small" compared to the problem size, as I'm sure you're aware.  From a software engineering point of view it's important that we properly handle this corner case of a large k_dim.  From an applied mathematics point of view it is probably more efficient to keep k_dim << max_it.

Best wishes,
Mark

**Comment by qiuchangkai** [comment]
Hi @mlstowell ,

I notice that you set up a pull request for `FGMRESSolver` already. Thanks for investigating this problem.

I prefer the option 3 for computing `Div(J)` too. The computational cost is not so important for me. I just want to maintain charge conservation and see how much divergence is in the solution field `E.`

Could you please show me some pseudo codes for option 3? I am still confused about how I can implement the code. This is the first time for me to know `L^2`, `H(div)` and `RT? mass matrix`.

Thanks for your help in advance.

Best,
Changkai

**Comment by mlstowell** [comment]
Hi, @qiuchangkai ,

I'd be happy to help.  `L^2` refers to a space of scalar-valued squared integrable functions.  These fields are discontinuous and do not have a natural type of derivative.  `RT` stands for Raviart-Thomas and refers to the researchers who first published a finite element space suitable for `H(Div)` fields.  `H(Div)` refers to the space of vector-valued fields which have square integrable divergence.

The code for option 3 would be something like this:
```
RT_FECollection RT_FEC(order-1, dim);
L2_FECollection L2_FEC(order-1, dim);
ParFiniteElementSpace RT_FESpace(&pmesh, &RT_FEC);
ParFiniteElementSpace L2_FESpace(&pmesh, &L2_FEC);
ParGridFunction j(&RT_FESpace);
ParGridFunction divj(&L2_FESpace);

ParMixedBilinearForm mSigma(&ND_FESpace, &RT_FESpace);
mSigma.AddDomainIntegrator(new VectorFEMassMatrix(sigma));
mSigma.Assemble();

ParBilinearForm m(&RT_FESpace);
m.AddDomainIntegrator(new VectorFEMassMatrix);
m.Assemble();

ParDiscreteLinearOperator div(&L2_FESpace, &RT_FESpace);
div.AddDomainInterpolator(new DivergenceInterpolator);
div.Assemble();

Vector sigmaE(mSigma.Height());
mSigma.Mult(e, sigmaE); // where 'e' is a ParGridFunction representation of the electric field, 'E'

HypreParMatrix M;
Vector SigmaE, J;
m.FormLinearSystem(ess_tdof_list, j, sigmaE, M, J, SigmaE);

// Setup a solver and compute J=M^{-1}SigmaE

m.RecoverFEMSolution(J, sigmaE, j); // Compute the H(Div) representation of current density in 'j'

div.Mult(j, divj);
```
I think that's everything although there may be some typos.

Best wishes,
Mark

**Comment by mlstowell** [comment]
Hi, @qiuchangkai ,

It looks like the changes to FGMRES have already been approved.  They should make their way into the master branch of MFEM in about one week, after a series of tests has been completed.  Thanks for pointing out this issue!

Best wishes,
Mark

**Comment by qiuchangkai** [comment]
Hi, @mlstowell ,

"FiniteElement::ProjectDiv(...) is not implemented for this element!" was reported when I run the following code:


	PetscInt order = 1;
	PetscInt dim = 3;
	RT_FECollection RT_FEC(order-1, dim);
	L2_FECollection L2_FEC(order-1, dim);
	ParFiniteElementSpace RT_FESpace(pmesh, &RT_FEC);
	ParFiniteElementSpace L2_FESpace(pmesh, &L2_FEC);
	ParGridFunction j_(&RT_FESpace);
	ParGridFunction divj(&L2_FESpace);

	Vector sigma;
	// set up sigma here
	PWConstCoefficient  beta_sigma(sigma);

	ParMixedBilinearForm mSigma(nd_fespace, &RT_FESpace);
	mSigma.AddDomainIntegrator(new VectorFEMassIntegrator(beta_sigma));
	mSigma.Assemble();

	ParBilinearForm m(&RT_FESpace);
	m.AddDomainIntegrator(new VectorFEMassIntegrator(beta_sigma));
	m.Assemble();

	ParDiscreteLinearOperator div(&L2_FESpace, &RT_FESpace);
	div.AddDomainInterpolator(new DivergenceInterpolator);
	div.Assemble();
// the error happens when trying to assemble the `div` operator

**Comment by mlstowell** [comment]
Ah, yes, I see a typo.  When declaring the `ParDiscreteLinearOperator` I reversed the finite element spaces.  The `RT` space should be first followed by the `L2` space.

**Comment by qiuchangkai** [comment]
Hi Mark, @mlstowell 

Now I understand the basic idea of option 3. For the calculation of `Div(sigma E)` , we first define a `ParMixedBilinearForm` `mSigma` for the operator `sigma E` to compute the current density `J` (normal continuous in `RT_FESpace` ) from the electric field `E` (tangential continuous in `ND_FESpace` ). Then a divergence operator `div` is setup for current density in `RT_FESpace`, to calculate the divergence of current density, resulting `divj`(discontinuous in `L2_FESpace`).

My input is only `PWConstCoefficient` `sigma` and `Vector` `e`. In the above code, `j` and `J` are both zero, so `divj` is also `zero`.  Are the `Vector` `sigmaE` and `SigmaE` not the actual current density? They have the same value in my code now. What are the variables `j` and `J` for?

Best,
Changkai

**Comment by mlstowell** [comment]
Hi, Changkai @qiuchangkai ,

There is one place in the code that I used a shortcut and just wrote a one line comment; the line mentioning `J=M^{-1}SigmaE`.  This was meant to be replaced with the solver of your choosing (I recommend PCG with a DiagScale preconditioner) I just didn't bother to write out the lines that setup that solver.  If you include this `J` won't be zero and the line with `RecoverFEMSolution` will set `j` appropriately.

The vectors `sigmaE` and `SigmaE` are certainly related to the current density but they are 'dual' vectors rather than 'primal' vectors.  In this context 'primal' vectors contain the finite element degrees of freedom i.e. the expansion coefficients in a finite element approximation of a field.  A 'dual' vector contains the field multiplied by each finite element basis function and integrated over the domain.  In MFEM 'dual' vectors are produced by `ParLinearForm` objects or they are the result of multiplying a `ParGridFunction` (a.k.a. a 'primal' vector) by a bilinear form.

The distinction between `sigmaE` and `SigmaE` is also not obvious.  They are both 'primal' vectors.  The lowercase vector, `sigmaE`, is a `ParGridFunction` which contains all of the degrees of freedom for each element in the local portion of the mesh.  The uppercase vector, `SigmaE`, contains a possibly reduced set of  degrees of freedom which are involved in the solution of the linear system.  Normally these two vectors are identical when codes are run in serial but in parallel `sigmaE` is larger because it contains copies of degrees of freedom which are shared between processors.  These vectors will also differ, even in serial, if your problem involves a non-conforming mesh (when using AMR on quadrilateral or hexahedral meshes), or if you use high order basis functions with static condensation or hybridization enabled.

Best wishes,
Mark

**Comment by qiuchangkai** [comment]
Hi Mark, @mlstowell ,

The option 3 is the Gauss divergence theorem.   What if I want to try option 2, i.e., to obtain the divergence at the vertex of the grids. I implement the following code:

	Vector sigma;
	// set up sigma here
	PWConstCoefficient  sigma_c(sigma);
	ParBilinearForm a(nd_fespace);
	a.AddDomainIntegrator(new VectorFEMassIntegrator(sigma_c));
	a.Assemble();
	HypreParMatrix mass;
	a.FormSystemMatrix(ess_tdof_list,mass);

	ParMixedBilinearForm div(nd_fespace, h1_fespace);
	ConstantCoefficient one(1.0e0);
	div.AddDomainIntegrator(new VectorFEWeakDivergenceIntegrator(one));
	div.Assemble();

	Vector e; // this is my input
	Vector j; // current density in H(curl) space
	Vector divj; // div in H1 space

	mass.Mult(e,j);
	div.Mult(current_density, divj);

Obviously I am wrong because there are no "2 mavtecs and **2 solves**" as you said. 

Any comments? Thank you very much.

Changkai

**Comment by tzanio** [comment]
The original question in this issue was addressed in #658.
